src/coreclr/vm/jithelpers.cpp

```diff
@@ -589,7 +621,13 @@ HCIMPLEND
 HCIMPL1_V(UINT64, JIT_Dbl2ULng, double val)
 {
     FCALL_CONTRACT;
+#if defined(TARGET_X86) || defined(TARGET_AMD64)
 
+    const double uint64_max_plus_1 = -2.0 * (double)INT64_MIN;
+    val = TruncateDouble(val);
```

Comment:
What is the perf impact of this change? Is this the best way to implement this helper with new behavior on x86/x64?

Reply:
What exactly do you mean by perf impact? Do you need the benchmarking results for the same with and without these changes on AVX512 and non AVX512 machine?

Reply:
Yes, we should have perf number for both AVX512 and non-AVX512 so that we understand the improvements and regressions.

Reply:
I will be providing perf numbers once we reach a final implementation. I hope that is fine.

Reply:
Here is an overview of the perf numbers with the current changes

### enabling AVX512F with the new changes
|              Method |      Mean |     Error |    StdDev | Median |    Min |       Max | Allocated |                                                                                                 
|-------------------- |----------:|----------:|----------:|-------:|-------:|----------:|----------:|                                                                                                 
| ConvDoubleToUlongBM | 0.0069 ns | 0.0089 ns | 0.0102 ns | 0.0 ns | 0.0 ns | 0.0325 ns |         - |

### disabling AVX512F with the new changes
|              Method |     Mean |    Error |   StdDev |   Median |      Min |      Max | Allocated |                                                                                                 
|-------------------- |---------:|---------:|---------:|---------:|---------:|---------:|----------:|                                                                                                 
| ConvDoubleToUlongBM | 9.816 ns | 1.124 ns | 1.294 ns | 10.32 ns | 5.703 ns | 10.40 ns |         - | 

### enabling avx512F without the new changes
|              Method |     Mean |    Error |   StdDev |   Median |      Min |      Max | Allocated |                                                                                                 
|-------------------- |---------:|---------:|---------:|---------:|---------:|---------:|----------:|                                                                                                 
| ConvDoubleToUlongBM | 368.7 ns | 88.13 ns | 101.5 ns | 441.0 ns | 232.0 ns | 448.1 ns |         - |

### disabling avx512F without the new changes
|              Method |     Mean |    Error |   StdDev |   Median |      Min |      Max | Allocated |
|-------------------- |---------:|---------:|---------:|---------:|---------:|---------:|----------:|
| ConvDoubleToUlongBM | 317.8 ns | 50.24 ns | 57.85 ns | 292.3 ns | 238.9 ns | 399.4 ns |         - |

We are seeing a significant performance increase on machine with AVX512

Reply:
Could you please share the source code for the microbenchmark?

Reply:
```
using System;
using BenchmarkDotNet.Attributes;
using MicroBenchmarks;

namespace System.MathBenchmarks
{
    public partial class Double
    {
        // Tests Math.Abs(double) over 5000 iterations for the domain -1, +1
        private static double val = 1.0;

        [Benchmark]
        public ulong ConvDoubleToUlongBM() => ConvDoubleToUlong(val);

        public static ulong ConvDoubleToUlong(double value)
        {
            return (ulong)value;
        }
    }
}
```

Reply:
The number you have shared do not look right. Double to ULong conversion should take about a nanosecond on current hardware.

`0.0069 ns` is way too low. It suggests that the operation was optimized out completely somehow.
`368.7 ns` and `317.8 ns` is way too high for any current hardware. `9.816 ns` is on the high side too.

For reference, this is what I am getting for the microbenchmark on .NET 8 Preview 5 on Surface Book 3 (Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz).

|              Method |     Mean |     Error |    StdDev |
|-------------------- |---------:|----------:|----------:|
| ConvDoubleToUlongBM | 1.202 ns | 0.0378 ns | 0.0354 ns |

Reply:
> Double to ULong conversion should take about a nanosecond on current hardware.

This doesn't seem right and I imagine both the above and the numbers you shared are running into the general issue that we're trying to measure something that executes in less time than the accuracy of the underlying hardware timer (which is at best 10-15ns, and more typically 100-300ns when considering OS API overhead).

Today, `double->ulong` conversion is always a helper fcall, a floating-point comparison, a branch, and a conversion instruction. The conversion instruction, even on the latest hardware takes 4-8 cycles and so we're looking at a bare minimum of 10 dependent cycles in the absolute ideal case and without considering the overhead of `call` or any prologue/epilogue. On a 1.3GHz computer, that'd be just shy of 10ns. Even on a fancier 4.5GHz machine we're looking at least at 2ns of execution time.

BDN has some known issues, which we regularly see in perf triage, around measuring functions that take such short amounts of time.

-----------------

Clang, GCC, and MSVC all do this roughly the same way, differing slightly in their algorithms.

For Clang it is branch free, while GCC branches. Both effectively do a signed conversion and conditionally subtract `0x43E0000000000000` (`9223372036854775808`). Neither of these platforms handle returning a sentinel value. For MSVC, it does some extra handling for values greater than or equal to `ulong.MaxValue`.

The rough pseudo-code is below.

### Clang
```csharp
long tmp1 = (long)(x);
long tmp2 = (long)(x - 9223372036854775808.0);
return (ulong)((tmp2 & (tmp1 >> 63)) | tmp1);
```

### GCC
```
if (x >= 9223372036854775808.0)
{
    long tmp1 = (long)(x - 9223372036854775808.0);
    return (long)(tmp1 ^ (1L << 63)); // btc
}
return (long)(x);
```

### MSVC
```
if (x < 9223372036854775808.0)
{
    return (long)(x);
}

x -= 9223372036854775808.0;

if (x >= 9223372036854775808.0)
{
    return (long)(x);
}

long tmp1 = (long)(x);
return (ulong)(tmp1 + 0x80000000_00000000);
```

-----------------

In our case, we want to treat anything negative as `0` and anything greater than or equal to `18446744073709551616` as `18446744073709551615`.

This means our code, at a minimum, needs to look something like:
```csharp
if (x < 0)
{
    return 0;
}

if (x < 9223372036854775808.0)
{
    return (long)(x);
}

x -= 9223372036854775808.0;

if (x < 9223372036854775808.0)
{
    return (long)(x);
}

return unchecked((ulong)(-1));
```

If we were to optimize this to be assembly code instead, we'd be looking at:
```asm
xorps    xmm1, xmm1    ;
maxsd    xmm0, xmm1    ; x = max(x, +0.0);             // Take +0.0 for NaN or if x is -0.0
cvtsd2si rcx,  xmm0    ; long tmp1 = (long)(x);        // Compliant conversion for [0, 9223372036854774784] and NaN, -9223372036854775808 for anything larger
subsd    xmm0, [cns]   ; x -= 9223372036854775808.0
cvtsd2si rax,  xmm0    ; long tmp2 = (long)(x);        // Compliant conversion, offset by `9223372036854775808` for [9223372036854775808, 18446744073709549568], -9223372036854775808 for anything larger and for NaN
mov      rdx,  rcx     ;
sar      rdx,  63      ; long tmp3 = tmp1 >> 63;       // 0 for [0, 9223372036854774784] and NaN, -1 for anything larger
and      rax,  rdx     ; 
sar      rax,  63      ; tmp2 = (tmp2 & tmp3) >> 63;   // 0 for [0, 9223372036854774784] and NaN, offset by `9223372036854775808` for [9223372036854775808, 18446744073709549568], -1 for anything larger
or       rax,  rcx     ; return (ulong)(tmp2 | tmp1);  // Compliant conversion for [0, 18446744073709549568] and NaN, ulong.MaxValue for anything larger
```

This would be 3 more instructions than Clang (`xorps`, maxsd`, and an additional `sar`). We can't really get better on pre-AVX-512 hardware without introducing additional branches/checks/handling. Even on `Haswell` (which is over 10 yo now) this is only 5 additional cycles to do correct saturation.

Reply:
> the numbers you shared are running into the general issue that we're trying to measure something that executes in less time than the accuracy of the underlying hardware timer

I do not think that it is the case. I have created a small standalone perf test to get Benchmark.NET out of the picture:
```csharp
using System.Diagnostics;
using System.Runtime.CompilerServices;

for (;;)
{
    var sw = new Stopwatch();
    sw.Start();
    for (int i = 0; i < 1_000_000_000; i++)
    {
        ConvDoubleToUlong(1.0);
    }
    Console.WriteLine(sw.ElapsedMilliseconds);
}

[MethodImplAttribute(MethodImplOptions.NoInlining)]
static ulong ConvDoubleToUlong(double value)
{
    return (ulong)value;
}
```
This loop takes about 2000ms on my machine that translates to 2ns per call. Benchmark.NET tries to subtract the cost of no-op call from the number that make the 1ns reported by Benchmark.NET pretty believable.

> we're looking at a bare minimum of 10 dependent cycles 

There is no hard dependency between different invocation of the call and that would force the execution to be serialized. Also, the whole loop is very small so it will easily fit into the microcode caches, etc. It is plausible that the hardware is very good at executing this microbenchmark in parallel firing on all cylinders.

Reply:
> There is no hard dependency between different invocation o

I was referring to the dependency between instructions. That is, for a single `(ulong)(someDouble)` there is no pipelining available for our current codegen.

It looks like the nuance here is coming from additional conversion cost that *can* show up for `cvttsd2si` for some, but not all, inputs and so the documented `4-8` cycles is more than it could be and it can take as few as 1 cycle based on what uProf is reporting.

Reply:
> It looks like the nuance here is coming from additional conversion cost that _can_ show up for `cvttsd2si` for some, but not all, inputs and so the documented `4-8` cycles is more than it could be and it can take as few as 1 cycle based on what uProf is reporting.

Just for my understanding, Here, "_The documented 4-8 cycles_" refer to the cycles needed for a single conversion(vcvtsd2si) instruction right?

Reply:
Right. They are documented (depending on if you're looking at AMD, Intel, or an external measurement like Agner's or uops.info) to be `<= 4` typically, but sometimes up to `8` depending on hardware in question.

Reply:
@tannergooding @jkotas 
We did a stopwatch method testing for these changes. They look consistent with what we would expect with the changes. Pasting the numbers below -->

<html>
<body>
<!--StartFragment--><span><span class="ui-provider fz b c d e f g h i j k l m n o p q r s t u v w x y z ab ac ae af ag ah ai aj ak" dir="ltr"><p><b>Ulong To Double (millisecond)</b></p><p></p><figure class="table ck-widget ck-widget_with-selection-handle ck-widget_selected" contenteditable="false"><span class="ck ck-widget__selection-handle"></span>

Diff, AVX512 disabled | Diff, AVX512 enabled | Base
-- | -- | --
1443 | 1152 | 1438
1465 | 1150 | 1436
1475 | 1152 | 1438
1478 | 1179 | 1438
1469 | 1175 | 1448

<span class="ck ck-reset_all ck-widget__type-around"><span class="ck ck-widget__type-around__fake-caret"></span></span></figure><p></p><br></span></span><!--EndFragment-->
</body>
</html>


<html>
<body>
<!--StartFragment--><span><span class="ui-provider fz b c d e f g h i j k l m n o p q r s t u v w x y z ab ac ae af ag ah ai aj ak" dir="ltr"><p><b>Ulong To Float (millisecond)</b></p><p></p><figure class="table ck-widget ck-widget_with-selection-handle ck-widget_selected" contenteditable="false"><span class="ck ck-widget__selection-handle"></span>

Diff | Diff, AVX512 | Base
-- | -- | --
1466 | 1176 | 1436
1463 | 1176 | 1438
1469 | 1179 | 1438
1467 | 1175 | 1439
1468 | 1174 | 1441

<span class="ck ck-reset_all ck-widget__type-around"><span class="ck ck-widget__type-around__fake-caret"></span></span></figure><p></p></span></span><!--EndFragment-->
</body>
</html>


**Source Code for Stopwatch Method testing**

```C#
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Runtime.Intrinsics.X86;
using System.Runtime.Intrinsics;
using System.Diagnostics;
public class Double
{
	private static int iter = 1_000_000_000;
	
	[MethodImplAttribute(MethodImplOptions.NoInlining)]
	public static ulong ConvDoubleToUlong(double dvalue)
	{
		return (ulong)dvalue;
	}
	
	[MethodImplAttribute(MethodImplOptions.NoInlining)]
	public static double ConvUlongToDouble(ulong dvalue)
	{
		return (double)dvalue;
	}
	
	[MethodImplAttribute(MethodImplOptions.NoInlining)]
	public static float ConvUlongToFloat(ulong dvalue)
	{
		return (float)dvalue;
	}
	
	public static void Main(string[] args)
	{
		Console.WriteLine("Double to ulong\n");
		Console.WriteLine(ConvDoubleToUlong(1.0));
		for (int q = 0;q < 5;q++)
		{
			var sw = new Stopwatch();
			sw.Start();
			for (ulong i = 0; i < 1_000_000_000; i++)
			{
				ConvDoubleToUlong(1.0);
			}
			Console.WriteLine(sw.ElapsedMilliseconds);
		}
		
		Console.WriteLine("ulong to double\n");
		for (int j = 0;j < 5;j++)
		{
			var sw = new Stopwatch();
			sw.Start();
			for (ulong i = 0; i < 1_000_000_000; i++)
			{
				ConvUlongToDouble(2);
			}
			Console.WriteLine(sw.ElapsedMilliseconds);
		}
		
		Console.WriteLine("ulong to float\n");
		for (int k = 0;k < 5;k++)
		{
			var sw = new Stopwatch();
			sw.Start();
			for (ulong i = 0; i < 1_000_000_000; i++)
			{
				ConvUlongToFloat(2);
			}
			Console.WriteLine(sw.ElapsedMilliseconds);
		}
	}
}

```

Also,

**We are narrowing down the scope of the PR to ulong -> float conversions.**

- The changes will b e updated soon.